### ISSUE TITLE
Add support for storing meta attributes as objects

### ISSUE THREAD

**Issue by paulwilde** [WordPress/gutenberg#6074]
I've developed a Map block type where you can plot out the markers which are plotted on the map. Each marker needs to store a few params each, the address, latitude and longitude. However, as meta attributes do not support using objects I need to use an ugly hack of storing this data as a string using a delimiter.

Here is how I'm storing the marker for attached screenshot:
`Mayfair, London, UK|51.5116269|-0.14780599999994593`

<img width="1497" alt="screen shot 2018-04-09 at 10 05 05" src="https://user-images.githubusercontent.com/704594/38489197-8c266f4e-3bdd-11e8-8306-ed7e065e0fdf.png">

Core does support storing/fetching meta objects using `update_post_meta`/`get_post_meta`, however it seems `register_meta` doesn't seem to support it, so that would need to be implemented first presumedly.

**Comment by gziolo** [comment]
Can you JSON serialize/deserilize them instead? It would work with both JS and PHP.

**Comment by youknowriad** [comment]
Related https://core.trac.wordpress.org/ticket/43392

**Comment by paulwilde** [comment]
> Can you JSON serialize/deserilize them instead? It would work with both JS and PHP.

Yeah, good idea. Totally didn't think about doing that - This actually fixes my use-case. 

It would be nice if Gutenberg supported running `JSON.stringify` and `JSON.decode` automatically if the attribute type is set as `JSON`.

**Comment by gziolo** [comment]
> Which is the best document to include some mention of JSON serialize/deserialize?

I would consider adding a note in https://github.com/WordPress/gutenberg/blob/master/docs/attributes.md.

> It would be nice if Gutenberg automatically ran `JSON.stringify` and `JSON.parse` automatically if the attribute type is set as JSON.

@aduth or @youknowriad, how complex it would be to add support for JSON type? It's still `string` internally for the `pgejs` parser if it cares at all.

**Comment by danielbachhuber** [comment]
> I would consider adding a note in https://github.com/WordPress/gutenberg/blob/master/docs/attributes.md.

Thanks!

> how complex it would be to add support for JSON type?

I'm 👎 on formally adding a `json` type. JSON is structured data masquerading as a string, and I suspect formally recognizing it will open a can of worms.

For now, I think it's better to let developers handle deeply structured data as they wish. If it becomes a common pattern and there's a more obvious abstraction, then we can pursue it at a later date.
